### DISCUSSION TITLE
nvPM EIn and energy forcing for changing fuel hydrogen content

### DISCUSSION THREAD

**Discussion by GunnarQuante** [contrailcirrus/pycontrails#118] [Q&A]
I am working on a study about the contrail climate impacts of different fuels. For this, I run Cocip over the same timeframe and fleet and vary the fuel's hydrogen content, from 13.8m-% to 15.3m-%. Even after a lot of testing, I still don't manage to replicate the results from the Teoh (2022) paper on targeted SAF use. In my case, e.g., the mean nvPM EIn decreases only by ~10% (for 13.8m-% to 15.3m-% hydrogen content) and the contrail energy forcing decreases by ~20%. While the latter is weather and fleet dependent, I would expect the change in nvPM EIn to be quite similar to Figure 1 of Teoh2022 (decrease by ~50%) 

![Figure 2023-12-05 084047](https://github.com/contrailcirrus/pycontrails/assets/141010511/bf4aafa1-5085-4f5e-8467-11d002c6123a)

![Validation_byTeoh2022](https://github.com/contrailcirrus/pycontrails/assets/141010511/ef2d227c-c25f-4210-a160-f40ebf0821fc)

Additionally, the increased hydrogen content also increases the fuel's energy content. Shouldn't this effect also reduce the fuel mass consumption for an identical fleet and weather situation? 

Here is a summary of the simulation results: 

          Day                     hydrogen content        mean_nvpm	        fuel use	             energy forcing
0  2019-01-01             13.80                            4.334682e+18	9.741743e+06	     6.638329e+15
1  2019-01-01             14.25                            4.097855e+18	9.741743e+06	     6.154426e+15
2  2019-01-01             14.55                            3.984812e+18	9.741743e+06	     5.840542e+15
3  2019-01-01             14.85                            3.896578e+18	9.741743e+06	     5.676103e+15
4  2019-01-01             15.30                            3.796649e+18	9.741743e+06	     5.521608e+15

I am not sure whether I'm missing something in my code or there is a bug in the pycontrails/Cocip code. A snippet of my code is attached - any help would be highly appreciated! 

[231205_GitHubSnippet.txt](https://github.com/contrailcirrus/pycontrails/files/13557576/231205_GitHubSnippet.txt)

**Comment by roger-teoh** [comment]
Hi @GunnarQuante , thank you for your question. It is great that you are trying to replicate our results over the North Atlantic. Can we schedule a time to jump on a call to discuss this further, so I can understand your approach in more detail (i.e., air traffic data, meteorology, etc.)?

**Reply by GunnarQuante** [reply]
Thank you, @roger-teoh for your support! Apparently using the pre-defined SAF-Blend instead of a custom fuel did the trick for most of the values: 

![Figure 2023-12-13 144402](https://github.com/contrailcirrus/pycontrails/assets/141010511/9ae3063b-d702-41b2-af2f-6a991a4c93ec)


What I still cannot make sense of is the fuel consumption. With an increasing hydrogen content also the fuel's energy content increases and thus the fuel consumption by mass should decrease. In my case, however it is still constant for a hydrogen content from 13.8m-% up to 15.3m-%. Here is a short example which should replicate the behavior (I'm using pycontrails 48.0 and python 3.10.11): 



    # import packages
    import pandas as pd
    from pycontrails_bada.bada_model import BADAFlight
    from pycontrails import Flight # , Aircraft
    from pycontrails.core.fleet import Fleet
    from pycontrails.core.fuel import JetA, SAFBlend
    from pycontrails import DiskCacheStore
    from pycontrails.datalib.ecmwf import ERA5
    from pycontrails.models.cocip import Cocip
    from pycontrails.models.humidity_scaling import ExponentialBoostLatitudeCorrectionHumidityScaling
    
    # demo synthetic flight
    flight_attrs = {
        "aircraft_type": "E190",
        # "wingspan": 48,
        # "n_engine": 2,
        "flight_id": "test",
        # set constants along flight path
        "true_airspeed": 226.099920796651,  # true airspeed, m/s
        # "thrust": 0.22,  # thrust_setting
        # "nvpm_ei_n": 1.897462e15,  # non-volatile emissions index
    }
    
    # Example flight
    df = pd.DataFrame()
    df["longitude"] = np.linspace(-25, -40, 100)
    df["latitude"] = np.linspace(34, 40, 100)
    df["altitude"] = np.linspace(10900, 10900, 100)
    df["engine_efficiency"] = np.linspace(0.34, 0.35, 100)
    df["fuel_flow"] = np.linspace(2.1, 2.4, 100)  # kg/s
    df["aircraft_mass"] = np.linspace(154445, 154345, 100)  # kg
    df["time"] = pd.date_range("2019-01-01T00:15:00", "2019-01-01T02:30:00", periods=100)
    
    flight = Flight(df, attrs=flight_attrs)
    
    flights = []
    for i in range(10):
        fl = flight.copy()
        fl.attrs.update(flight_id=f"test-{i}")
        fl.update(latitude=flight["latitude"] + i)
        fl.update(longitude=flight["longitude"] + i)
        flights.append(fl)
        
    # either load Weather data from cache or download from ECMWF
    time_format = "%Y-%m-%d %H:%M"
    time = (
        df["time"].iloc[0],
        df["time"].iloc[-1] + pd.Timedelta("10H"),
    )
    
    pressure_levels = [
        400,
        350,
        300,
        250,
        225,
        200,
        175,
        150,
    ]
    
    year = '2019'
    cache = 'C:\\Users\\Gunnar\\Desktop\\cache\\'
    savedir = cache + year
    
    cache_path = savedir + '\\ECMWF'
    disk_cache = DiskCacheStore(cache_dir=cache_path, allow_clear=True)
    era5pl = ERA5(time=time, variables=Cocip.met_variables, pressure_levels=pressure_levels, cachestore=disk_cache)
    era5sl = ERA5(time=time, variables=Cocip.rad_variables, cachestore=disk_cache)
    
    met = era5pl.open_metdataset()
    rad = era5sl.open_metdataset()
    
    # Define list of fuel types
    blends = [0, 30, 60, 100]
    fuel = []
    for blend_ratio in blends:
        fuel.append(SAFBlend(blend_ratio))
        print(SAFBlend(blend_ratio).hydrogen_content)
    hydrogen_cont = [f.hydrogen_content for f in fuel]
    
    params = {
        "process_emissions": True,
        "verbose_outputs": False,
        "humidity_scaling": ExponentialBoostLatitudeCorrectionHumidityScaling(met=met),
        "radiative_heating_effects": True,
    }
    
    bada_model = BADAFlight(met=None, bada_priority=3)
    
    ef = []
    nvPM = []
    fuel_flow = []
    aircraft_mass = []
    fuel_burn = []
    
    # Run Cocip for varying fuel types and the same fleet and same weather
    for fuel1 in fuel: 
        # flights2 = []
        for flight in flights: 
            flight.fuel = fuel1
            # flights2.append(flight)
        Fleet = Fleet.from_seq(flights)
        print(Fleet.fuel.fuel_name, 'before CoCiP, with hydrogen content', round(Fleet.fuel.hydrogen_content, 2), 'and LHV', round(Fleet.fuel.q_fuel, 0))
    
        cocip = Cocip(met=met, rad=rad, params=params, aircraft_performance=bada_model) # 
        output_flights = []
        output_flights = cocip.eval(source=Fleet)
        output_flights = output_flights.to_flight_list()
        flight = output_flights[-1]
        print(Fleet.fuel.fuel_name, 'after CoCiP, with hydrogen content', round(Fleet.fuel.hydrogen_content, 2), 'and LHV', round(Fleet.fuel.q_fuel, 0))
        ef.append(np.nansum([flight.data['ef'] for flight in output_flights]))
        nvPM.append(np.nansum([flight.data['nvpm_ei_n'] for flight in output_flights]))
        fuel_flow.append(np.nansum([flight.data['fuel_flow'] for flight in output_flights]))
        fuel_burn.append(np.nansum([flight.data['fuel_burn'] for flight in output_flights]))
        aircraft_mass.append(np.nanmean([flight.data['aircraft_mass'] for flight in output_flights]))
    
    print('EF:', ef, '\n', 'nvPM:', nvPM, '\n', 'fuel_flow:', fuel_flow, '\n', 'fuel_burn', fuel_burn, '\n', 'aircraft_mass:', aircraft_mass)

**Reply by GunnarQuante** [reply]
I think I found a solution for the fuel consumption: From how I understand the code, when using the bada performance model, the fuel consumption is calculated via the "fuel_mass_flow_rate" function in bada3.py (bada4.py). E.g., for jet aircraft it uses the BADA data to retrieve the thrust specific fuel consumption and multiplies it with the previously calculated thrust force. 

It seems to me that as of now the "fuel_mass_flow_rate" function does not take into account varying fuel energy contents. I suppose an adjustment factor (e.g. adjustment = JetA.q_fuel/SAFBlend.q_fuel) would allow for taking this effect into account. I added it to my code and this yields seemingly consistent results (e.g. a reduction of 2.5 % in the fleet's fuel consumption for an energy content increase by 2.5%).
